gossip/gossip.go

```diff
@@ -0,0 +1,123 @@
+// Copyright (C) 2019-2023, Ava Labs, Inc. All rights reserved.
+// See the file LICENSE for licensing terms.
+
+package gossip
+
+import (
+	"context"
+	"sync"
+	"time"
+
+	"github.com/ethereum/go-ethereum/log"
+
+	"github.com/ava-labs/avalanchego/codec"
+	"github.com/ava-labs/avalanchego/ids"
+	"github.com/ava-labs/avalanchego/network/p2p"
+)
+
+// GossipableAny exists to help create non-nil pointers to a concrete Gossipable
+type GossipableAny[T any] interface {
+	*T
+	Gossipable
+}
+
+type Config struct {
+	Frequency time.Duration
+	PollSize  int
+}
+
+func NewGossiper[T any, U GossipableAny[T]](
+	config Config,
+	set Set[U],
+	client *p2p.Client,
+	codec codec.Manager,
+	codecVersion uint16,
+) *Gossiper[T, U] {
+	return &Gossiper[T, U]{
+		config:       config,
+		set:          set,
+		client:       client,
+		codec:        codec,
+		codecVersion: codecVersion,
+	}
+}
+
+type Gossiper[T any, U GossipableAny[T]] struct {
+	config       Config
+	set          Set[U]
+	client       *p2p.Client
+	codec        codec.Manager
+	codecVersion uint16
+}
+
+func (g *Gossiper[T, U]) Gossip(shutdownChan chan struct{}, shutdownWg *sync.WaitGroup) {
+	gossipTicker := time.NewTicker(g.config.Frequency)
+	defer func() {
+		gossipTicker.Stop()
+		shutdownWg.Done()
+	}()
+
+	for {
+		select {
+		case <-gossipTicker.C:
+			if err := g.gossip(); err != nil {
+				log.Warn("failed to gossip", "error", err)
+			}
+		case <-shutdownChan:
+			log.Debug("shutting down gossip")
+			return
+		}
+	}
+}
+
+func (g *Gossiper[T, U]) gossip() error {
+	filter := g.set.GetFilter()
+	bloomBytes, err := filter.Bloom.MarshalBinary()
+	if err != nil {
+		return err
+	}
+
+	request := PullGossipRequest{
+		FilterBytes: bloomBytes,
+		SaltBytes:   filter.Salt,
+	}
+	msgBytes, err := g.codec.Marshal(g.codecVersion, request)
+	if err != nil {
+		return err
+	}
+
+	for i := 0; i < g.config.PollSize; i++ {
+		if err := g.client.AppRequestAny(context.TODO(), msgBytes, g.handleResponse); err != nil {
```

Comment:
Do we ensure we don't send to the same node twice in this loop?

Reply:
No we don't. We'd have to pass in an int to `AppRequestAny` to specify the amount of nodes to make the requests to. I can put up an upstream PR to `avalanchego` to do this.

Reply:
I also don't know if we have to block on this, we could also consider doing this iteratively in a follow-up as well.

Reply:
https://github.com/ava-labs/avalanchego/issues/1868